File: src/main/kotlin/com/inventory/horse/entity/Horse.kt
package com.inventory.horse.entity

import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.JoinColumn
import jakarta.persistence.ManyToOne
import jakarta.persistence.Table
import java.math.BigDecimal

@Entity
@Table(name = "horses")
data class Horse(
    @Id @GeneratedValue(strategy = GenerationType.IDENTITY)
    val id: Long = 0,
    @Column
    val tagged: Boolean,
    @ManyToOne
    @JoinColumn(name = "manufacturer_id", nullable = false)
    val manufacturer: Manufacturer,
    @ManyToOne
    @JoinColumn(name = "mold_id", nullable = false)
    val mold: Mold,
    @ManyToOne
    @JoinColumn(name = "scale_id", nullable = false)
    val scale: Scale,
    @ManyToOne
    @JoinColumn(name = "model_id", nullable = false)
    val model: Model,
    @ManyToOne
    @JoinColumn(name = "breed_id", nullable = false)
    val breed: Breed,
    @ManyToOne
    @JoinColumn(name = "breed_type_id", nullable = false)
    val breedType: BreedType,
    @ManyToOne
    @JoinColumn(name = "color_id", nullable = false)
    val color: Color,
    @ManyToOne
    @JoinColumn(name = "pattern_id", nullable = false)
    val pattern: Pattern,
    @ManyToOne
    @JoinColumn(name = "gender_id", nullable = false)
    val gender: Gender,
    @ManyToOne
    @JoinColumn(name = "condition_id", nullable = false)
    val condition: Condition,
    @ManyToOne
    @JoinColumn(name = "location", nullable = false)
    val location: Location,
    @Column(name = "purchase_price", precision = 10, scale = 2, nullable = false)
    val purchasePrice: BigDecimal,
    @Column(name = "sell_price", precision = 10, scale = 2, nullable = false)
    val sellPrice: BigDecimal,
    @Column(name = "nan_qualified", nullable = false)
    val nanQualified: Boolean,
    @Column(name = "first_place", nullable = false)
    val firstPlace: Int,
    @Column(name = "second_place", nullable = false)
    val secondPlace: Int,
    @Column(name = "third_place", nullable = false)
    val thirdPlace: Int,
    @Column(name = "fourth_place", nullable = false)
    val fourthPlace: Int,
    @Column(name = "fifth_place", nullable = false)
    val fifthPlace: Int,
    @Column
    val showName: String,
    @Column
    val officePony: String?,
) {
    constructor() : this(
        0,
        false,
        Manufacturer(),
        Mold(),
        Scale(),
        Model(),
        Breed(),
        BreedType(),
        Color(),
        Pattern(),
        Gender(),
        Condition(),
        Location(),
        BigDecimal("0.00"),
        BigDecimal("0.00"),
        false,
        0,
        0,
        0,
        0,
        0,
        "",
        "",
    )
}

File: frontend/src/types/Horse.ts
// src/types/Horse.ts
export interface NamedEntity {
  id: number;
  name: string;
}

export interface Manufacturer extends NamedEntity {}

export interface Mold extends NamedEntity {
  manufacturer: Manufacturer;
}

export interface RunType extends NamedEntity {}

export interface Finish extends NamedEntity {}

export interface Scale extends NamedEntity {}

export interface Model extends NamedEntity {
  mold: Mold;
  runType: RunType;
  finish: Finish;
  scale: Scale;
}

export interface Horse {
  id: number;
  showName: string;
  tagged: boolean;
  manufacturer: Manufacturer;
  mold: Mold;
  scale: Scale;
  model: Model;
  breed: NamedEntity;
  breedType: NamedEntity;
  color: NamedEntity;
  pattern: NamedEntity;
  gender: NamedEntity;
  condition: NamedEntity;
  location: NamedEntity;
  purchasePrice: number;
  sellPrice: number;
  nanQualified: boolean;
  firstPlace: number;
  secondPlace: number;
  thirdPlace: number;
  fourthPlace: number;
  fifthPlace: number;
  officePony?: string | null;
}

export interface HorseRequest {
  tagged: boolean;
  manufacturerId: number;
  moldId: number;
  scaleId: number;
  modelId: number;
  breedId: number;
  breedTypeId: number;
  colorId: number;
  patternId: number;
  genderId: number;
  conditionId: number;
  locationId: number;
  purchasePrice: number;
  sellPrice: number;
  nanQualified: boolean;
  firstPlace: number;
  secondPlace: number;
  thirdPlace: number;
  fourthPlace: number;
  fifthPlace: number;
  showName: string;
  officePony?: string | null;
}
